tmt/steps/prepare/reboot.py

```diff
@@ -0,0 +1,103 @@
+import dataclasses
+import os
+from typing import Any, List, Optional, cast
+
+import click
+import pkg_resources
+
+import tmt
+import tmt.steps.prepare
+import tmt.utils
+from tmt.steps.common import RebootCommon
+from tmt.steps.execute import TMT_REBOOT_SCRIPT
+from tmt.steps.provision import Guest
+
+# Scripts source directory
+SCRIPTS_SRC_DIR = pkg_resources.resource_filename(
+    'tmt', 'steps/execute/scripts')
+
+SCRIPTS = (
+    TMT_REBOOT_SCRIPT,
+    )
+
+# TODO: remove `ignore` with follow-imports enablement
+
+
+@dataclasses.dataclass
+class PrepareRebootData(tmt.steps.prepare.PrepareStepData):
+    script: List[str] = dataclasses.field(default_factory=list)
+
+    _normalize_script = tmt.utils.NormalizeKeysMixin._normalize_string_list
+
+
+@tmt.steps.provides_method('reboot')
+class PrepareReboot(
+        tmt.steps.prepare.PreparePlugin,
+        RebootCommon
+        ):
+    """
+    Reboot system via provided script
+    Example config:
+    prepare:
+        how: reboot
+        script: ./reboot-script
+    """
+
+    _data_class = PrepareRebootData  # type: ignore[assignment]
+
+    @classmethod
+    def options(cls, how: Optional[str] = None) -> Any:
+        """ Prepare command line options """
+        return cast(List[tmt.options.ClickOptionDecoratorType], [
+            click.option(
+                '-n', '--name', metavar='NAME',
+                help='Set name of the reboot step.'),
+            click.option(
+                '-n', '--script', metavar='NAME',
+                help='Set path to the reboot script.')
+            ]) + super().options(how)
+
+    def default(self, option: str, default: Optional[Any] = None) -> Any:
+        """ Return default data for given option """
+        if option == 'name':
+            return "Reboot script name"
+        if option == 'reboot':
+            return "reboot-script"
+        return default
+
+    def go(self, guest: Guest) -> None:
+        """ Prepare the guests """
+        super().go(guest)
+
+        # Prepare scripts, except localhost guest
+        self.scripts = SCRIPTS
+        self.debug(f"Preparing reboot script: {self.scripts}")
+        if not guest.localhost:
+            self.prepare_scripts(guest)
```

Comment:
Ha, this one seems to come from execute plugin - I'd argue it's now obvious other steps are interested in preparing scripts, therefore we should move this method outside of the execute plugin, e.g. by making it a function, instead of a method, accepting `scripts` to upload. Or moving it to `Guest` class, since it's preparing scripts on a particular guest. In any case, its use of `self.scripts` should go away, `scripts` as a parameter is a much better way.

Reply:
I'd vote for including it in this PR, to make the resulting implementation cleaner. Your patch is introducing the reboot capability to other steps, to me that sounds like a good point for exactly this kind of refactoring.

Reply:
OK, agreed, so I will include the changes here. Let me think about it a but :smirk: